new "type system concepts" section (#1743)
python/typing: docs/spec/concepts.rst

every operation will be checked, but certain primitive operations in the
language such as attribute access or arithmetic are. Python is a
dynamically-typed language.

**Gradual typing** is a way to combine static and dynamic typing.
Type-annotated Python allows opting in to static type checking at a fine level
of granularity, so that some type errors can be caught statically, before
running the program. Variables, parameters, and returns can optionally be given
a static type annotation. Even within the type of a single data structure,
static type checking is opt-in and granular. For example, a dictionary can be
annotated to have static checking of the key type but only dynamic checking of
the value type.

In gradual typing, a special "unknown" or "dynamic" type is used in the static
type system for names or expressions whose type is not known statically.  In
Python, this type is spelled :ref:`Any`. Because :ref:`Any` indicates a
statically-unknown type, the static type checker can't check type correctness
of operations on expressions typed as :ref:`Any`. These operations are
dynamically checked, via the Python runtime's usual dynamic checking.

> The Python type system also uses ``...`` within :ref:`callable` and :ref:`Tuple

Author: Fixed. Thanks for the reminder that only `tuple[Any, ...]` is a gradual form. I updated the wording here to be a bit more careful so that it neither specifies this (that should be left for the tuple chapter) nor contradicts it.

In the remainder of this specification, we will usually prefer the term
**assignable to** over "consistent subtype of". The two are synonymous, but
"assignable to" is shorter, and may communicate a clearer intuition to many
readers.

For example, ``Any`` is assignable to ``int``, because ``int`` is a
materialization of ``Any``, and ``int`` is a subtype of ``int``. The same
materialization also gives that ``int`` is assignable to ``Any``.

The assignable-to relation is not generally symmetric, though. If ``B`` is a
subtype of ``A``, then ``tuple[Any, B]`` is assignable to ``tuple[int, A]``,
because ``tuple[Any, B]`` can materialize to ``tuple[int, B]``, which is a
subtype of ``tuple[int, A]``. But ``tuple[int, A]`` is not assignable to
``tuple[Any, B]``.

Attributes and methods
----------------------

In Python, we can do more with objects at runtime than just assign them to
names, pass them to functions, or return them from functions. We can also
get/set attributes and call methods.

In the Python object model, the operations that can be performed on a value all
de-sugar to method calls. For example, ``a + b`` is syntactic sugar for either
``a.__add__(b)`` or ``b.__radd__(a)``.

> For a static type checker, accessing ``a.foo`` is a type error unless all

Reviewer: This statement isn't strictly true if some of the objects support `__getattr__` (or set / delete depending on the context). I'm not sure if it's worth muddying the description with this caveat though. Just wanted to point it out in case you want to do something with it.

Author: Thanks! Added a parenthetical about `__getattr__` to avoid misinterpretation.

I think it would be good for the spec to have a deeper discussion of attribute getters and setters and their types (and how this works with gradual and set-theoretic types), but this will be a quite nuanced topic, so I'd rather tackle it in a later PR.

Reviewer: Yes, I agree this deserves its own section.

FWIW, here are some [notes](https://github.com/erictraut/mypy_issues/blob/main/tc_topics.md) that I've been maintaining for myself about the typing spec and conformance tests. In these notes (which are admittedly rough and use a shorthand that may not make sense to others), I list some ideas for additional chapters and topics that I think are missing from the current spec.

possible objects in the set represented by the type of ``a`` have the ``foo``
attribute.

If all objects in the set represented by the fully static type ``A`` have a
``foo`` attribute, we can say that the type ``A`` has the ``foo`` attribute.

If the type ``A`` of ``a`` in ``a.foo`` is a gradual type, it may not represent
a single set of objects. In this case, ``a.foo`` is not a type error if and

> only if there exists a materialization of ``A`` which has the ``foo``

Reviewer: Or has a `__[get|set|del]attr__` method. Once again, not sure this is worth adding.

and verifies that their uses obey the typing rules. Normally, a program that is
not well-typed (i.e., one that contains a type error) will not run. Java and
C++ are examples of statically-typed object-oriented languages.

A **dynamically-typed** programming language does not run a type-checker before
running a program. Instead, it checks the types of values before performing
operations on them at runtime. This is not to say that the language is
"untyped". Values at runtime have a type and their uses obey typing rules. Not
every operation will be checked, but certain primitive operations in the
language such as attribute access or arithmetic are. Python is a
dynamically-typed language.

**Gradual typing** is a way to combine static and dynamic typing.
Type-annotated Python allows opting in to static type checking at a fine level
of granularity, so that some type errors can be caught statically, before
running the program. Variables, parameters, and returns can optionally be given
a static type annotation. Even within the type of a single data structure,
static type checking is opt-in and granular. For example, a dictionary can be

> annotated to have static checking of the key type but only dynamic checking of

Author: Taking the first suggestion ("enable" instead of "have") but not the second: to me "retain" wrongly suggests that otherwise we'd be giving up dynamic checking, which isn't actually the case in Python (as you highlight in a later comment.)

> annotated to have static checking of the key type but only dynamic checking of

Reviewer 2: I see, makes sense. I think I interpreted "dynamic typing" here to refer to the way the static type checker was meant to see the types (i.e., they should interpret anything unannotated as being of type `Any`). Perhaps something like this might be clearer:

```suggestion
annotated to enable static checking of the key type but only have dynamic runtime checking of
```

the value type.

In gradual typing, a special "unknown" or "dynamic" type is used in the static
type system for names or expressions whose type is not known statically.  In
Python, this type is spelled :ref:`Any`. Because :ref:`Any` indicates a
statically-unknown type, the static type checker can't check type correctness
of operations on expressions typed as :ref:`Any`. These operations are

> dynamically checked, via the Python runtime's usual dynamic checking.

Reviewer 2: > I think adding the word "still" to this sentence can help address your concern:
> 
> > These operations are _still_ dynamically checked, via the Python runtime's usual dynamic checking.

Yes, I think that would help!

The Python type system also uses ``...`` within :ref:`Callable` types and

> within ``tuple[Any, ...]`` (see :ref:`tuples`) to indicate a statically-unknown

Author: I agree with you that it's technically maybe not necessary to mention `tuple[Any, ...]` here, given that the `Any` would already make it a gradual type. I thought about this when writing it. But I do think it's worth mentioning here, because the `Any` and the `...` both give a graduality to `tuple[Any, ...]` along different axes, so "the ... in `tuple[Any, ...]`" really is a distinct gradual form from the `Any`. (Even though, confusingly, it is not a gradual form in e.g. `tuple[int, ...]`.)

Ultimately I feel like the current wording here is the most accurate choice. It's unfortunate that `tuple[Any, ...]` is a special case, but we need to leave it to the tuple docs to clarify that.

component of a type. The detailed rules for these usages are discussed in their
respective sections of the specification.

This specification describes a gradual type system for Python.

Fully static and gradual types
------------------------------

We will refer to types that do not contain :ref:`Any` (or the ``...`` in
``Callable`` or ``tuple[Any, ...]``) as a sub-part as **fully static types**.

A **gradual type** can be a fully static type, :ref:`Any` itself, or a type
that contains :ref:`Any` (or the ``...`` in ``Callable`` or ``tuple[Any,
...]``) as a sub-part.

Fully static types
~~~~~~~~~~~~~~~~~~

A fully static type denotes a set of potential runtime values. For instance,
the fully static type ``object`` is the set of all Python objects. The fully
static type ``bool`` is the set of values ``{ True, False }``. The fully static
type ``str`` is the set of all Python strings; more precisely, the set of all
Python objects whose runtime type (``__class__`` attribute) is either ``str``
or a class that inherits directly or indirectly from ``str``. A :ref:`Protocol
<Protocols>` denotes the set of all objects which share a certain set of

> attributes and/or methods.

Author: This is an interesting observation, but I think the distinction would be lost on most readers unless we explicitly highlighted it, and explicitly highlighting it would be an unnecessary digression here, I think. So I'm inclined not to extend this paragraph with an additional example.

Gradual types
~~~~~~~~~~~~~

:ref:`Any` represents an unknown static type. It denotes some unknown set of

> values.

Author: We said above "A fully static type denotes a set of potential runtime values." How many of those descriptors do you think it is necessary to repeat with each usage? What other interpretation of the term "value" are you worried about this being confused with?

> values.

Reviewer 2: I don't think there's any potential for confusion here, and I don't feel strongly about this one! To me this felt like this added some amount of symmetry between the two adjacent sentences; it felt nice for "unknown static type" to be balanced by "unknown [set of] runtime values" in the second sentence. Just incidental wordsmithing.

that operations on the expression are valid for all Python objects, or else
emit a static type error. This allows very few operations! For example, if
``x`` is typed as ``object``, ``x.foo`` should be a static type error because
not all Python objects have an attribute ``foo``.

An expression typed as :ref:`Any`, on the other hand, should be assumed to have
_some_ specific static type, but _which_ static type is not known. A static
type checker should not emit static type errors on an expression or statement
if :ref:`Any` might represent a static type which would avoid the error. (This
intuition is made more precise below, in our definitions of materialization and
assignability.)

Similarly, a gradual type such as ``tuple[int, Any]`` (see :ref:`tuples`) or
``int | Any`` (see :ref:`union-types`) does not represent a single set of
Python objects; rather, it represents a (bounded) range of possible sets of
values.

In the same way that ``Any`` does not represent "the set of all Python objects"
but rather "an unknown set of objects", ``tuple[int, Any]`` does not represent
"the set of all length-two tuples whose first element is an integer." That is a

> fully static type, spelled ``tuple[int, object]``.  In contrast, ``tuple[int,

Author: Can you say more about why you consider this an improvement?

> fully static type, spelled ``tuple[int, object]``.  In contrast, ``tuple[int,

Reviewer 2: A type is an abstract concept; it feels odd to me to say that you can "spell" an abstract concept. Words are spelled on a page; syntax is spelled on a page; abstract concepts are not spelled on a page. The specific notation we use to describe or express an abstract concept might be spelled on a page, but that is not the abstract concept itself that is being spelled.

> fully static type, spelled ``tuple[int, object]``.  In contrast, ``tuple[int,

Author: Hmm. That makes sense, but I also think this usage of "spelled" to mean "this is how you express this concept in syntax" is not uncommon. Above we say that the dynamic type in Python is "spelled" `Any` -- did that phrase also read oddly to you?

> fully static type, spelled ``tuple[int, object]``.  In contrast, ``tuple[int,

Reviewer 2: It does a _little_, yes, though it doesn't feel quite as bad to me as this case (and I'm struggling to come up with a good reason why :-)

Obviously not the biggest issue in the world either way!

Any]`` represents some unknown set of tuple values; it might be the set of all
tuples of two integers, or the set of all tuples of an integer and a string, or
some other set of tuple values.

In practice, this difference is seen (for example) in the fact that we can
assign an expression of type ``tuple[int, Any]`` to a target typed as
``tuple[int, int]``, whereas assigning ``tuple[int, object]`` to ``tuple[int,
int]`` is a static type error. (Again, we formalize this distinction in the
below definitions of materialization and assignability.)

In the same way that the fully static type ``object`` is the upper bound for
the possible sets of values represented by ``Any``, the fully static type
``tuple[int, object]`` is the upper bound for the possible sets of values

> represented by ``tuple[int, Any]``.

Author: There is an inherent tension here in wanting to use examples to clarify the concepts, while some of the examples themselves implicitly depend on more advanced concepts (like the idea of a generic type in the first place!) that we aren't ready to dive into yet.

That said, in this particular section I don't think there is any implicit assumption about variance. (There definitely is later on in the section on consistent subtyping.) I think everything said here would also be true of list?

> represented by ``tuple[int, Any]``.

Reviewer 2: I'm not sure `list[object]` is the upper bound for the possible sets of values represented by `list[Any]`, is it? `list[object]` is not a supertype of `list[int]`, but `list[Any]` could be materialised as `list[int]`

> represented by ``tuple[int, Any]``.

Author: Regarding the case below where we do assume covariance of tuple, in discussing consistent subtyping: I think in practice this is OK because covariance is definitely most people's natural intuition ("yes, a list of strings is a list of objects!"). It's invariant generic containers that people find unintuitive and need explanation for. So I think it works fine to use tuples in examples in this doc, with the assumption of covariance, and leave further discussion of variance to the generics doc.

> represented by ``tuple[int, Any]``.

Reviewer 2: Definitely agree that a detailed discussion of variance is out of scope here.

> represented by ``tuple[int, Any]``.

Author: Ah, of course you're right, `list[object]` is not an upper bound for `list[Any]`, exactly as you say. There is a covariance assumption here as well.

The gradual guarantee
~~~~~~~~~~~~~~~~~~~~~

:ref:`Any` allows gradually adding static types to a dynamically-typed program.
For a fully dynamically-typed program, a static checker has the type :ref:`Any`

> for all expressions, and should emit no errors. Adding static type annotations

Author: I don't like "infers", because in a fully dynamically typed program, the static type checker shouldn't be inferring any static types either. Assigns might be good here.

> for all expressions, and should emit no errors. Adding static type annotations

Author: To some extent the statement here is tautological. If a static type checker begins to infer some non-Any types (and clearly there are cases where it would be safe for it to do so!) then the program is, by definition, no longer "fully dynamically typed"! But I think the tautological statement is still useful, demonstrating one extreme end of the spectrum.

I agree that in this case we should maybe clarify that a fully dynamically typed program can become a bit more statically typed without even adding annotations. I'll think about whether I can slip that in without making things too wordy.

> to the program (making the program more statically typed) may result in static

Author: I think it is useful, because it calls back to our earlier definitions of "statically typed" and "dynamically typed."

> to the program (making the program more statically typed) may result in static

Reviewer 2: It might read a bit better if you rephrased it to "Adding type annotations" rather than "adding _static_ type annotations". That would mean you wouldn't have "static" twice in the same sentence (and I don't think there's any other kind of type annotation in Python, other than static type annotations)

type errors, if the program is not correct or if the static type annotations
aren't able to fully represent the runtime types. Removing static type
annotations (making the program more dynamic) should not result in additional
static type errors. This is often referred to as the **gradual guarantee**.

In Python's type system, we don't take the gradual guarantee as a strict
requirement, but it's a useful guideline.

Subtype, supertype, and type equivalence
----------------------------------------

A fully static type ``B`` is a **subtype** of another fully static type ``A``
if and only if the set of values represented by ``B`` is a subset of the set of
values represented by ``A``. Because the subset relation on sets is transitive

> and reflexive, the subtype relation is also transitive (if ``C`` is a subtype

Author: I feel like this edit implies that there is also a subtype relation on not-fully-static types, which is not true (and it's important that it's not true.)

> and reflexive, the subtype relation is also transitive (if ``C`` is a subtype

Reviewer 2: Right. This was another wordsmithing suggestion, for symmetry between the two halves of the sentence ("the subset relation on sets" becomes balanced by "the subtype relation on static types"). Maybe this?

```suggestion
and reflexive, the subtype relation on types is also transitive (if ``C`` is a subtype
```

> and reflexive, the subtype relation is also transitive (if ``C`` is a subtype

Author: Ah, I see what you're going for. But I feel like "subtype relation on types" also risks implying that it is a relation on all types, including gradual ones.

I think this might be a case where the sentence symmetry is not as important as ensuring we don't imply anything inaccurate by adding unnecessary qualifiers.

> and reflexive, the subtype relation is also transitive (if ``C`` is a subtype

Reviewer 2: Makes sense.

if the set of values represented by ``A`` is a superset of the values
represented by ``B``. The supertype relation is also transitive and reflexive.

We also define an **equivalence** relation on fully static types: the types
``A`` and ``B`` are equivalent (or "the same type") if and only if ``A`` is a
subtype of ``B`` and ``B`` is a subtype of ``A``. This means that the set of
values represented by ``A`` is both a superset and a subset of the values
represented by ``B``, meaning ``A`` and ``B`` must represent the same set of
values.

We may describe a type ``B`` as "narrower" than a type ``A`` (or as a "proper
subtype" of ``A``) if ``B`` is a subtype of ``A`` and ``B`` is not equivalent
to ``A``. In the same scenario we can describe the type ``A`` as "wider" than
``B``, or a "proper supertype" of ``B``.

Materialization
---------------

Since :ref:`Any` represents an unknown static type, it does not represent any
known single set of values (it represents an unknown set of values.) Thus it is
not in the domain of the subtype, supertype, or equivalence relations on static
types described above.

To relate gradual types more generally, we define a **materialization**

> relation. The intuition for materialization is that it transforms a "more

Author: Yes, good edit, but sorry, going to stay consistent with American spelling 😆

> relation. The intuition for materialization is that it transforms a "more

Reviewer 2: dang

dynamic" type to a "more static" type. Given a gradual type ``A``, if we

> replace zero or more occurrences of ``Any`` in ``A`` with some gradual type

Author: Why is passive voice preferable to "we"?

> replace zero or more occurrences of ``Any`` in ``A`` with some gradual type

Reviewer 2: Who is "we" here? ;)

The general tone of this article is quite formal, and the general advice given by style guides is to avoid first-person pronouns in this kind of formal writing. To me this wouldn't be out of place in a tutorial, or some other kind of article that addresses the reader constantly. In a formal specification that aims to describe abstract concepts in an objective way, using "we" here feels out of place to me.

> replace zero or more occurrences of ``Any`` in ``A`` with some gradual type

Author: There are many, many similar uses of "we" in this article ("We will refer...", "we can assign", "we formalize...", "we don't take" all occur within the first twenty lines, where I stopped my search.) I feel like this is quite common in presentations of systems similar to this one.

> replace zero or more occurrences of ``Any`` in ``A`` with some gradual type

Reviewer 2: Alright, fair enough :)

(which can be different for each occurrence of ``Any``), the resulting gradual
type ``B`` is a materialization of ``A``.

For instance, ``tuple[int, str]`` (a fully static type) and ``tuple[Any, str]``
(a gradual type) are both materializations of ``tuple[Any, Any]``. ``tuple[int,
str]`` is also a materialization of ``tuple[Any, str]``.

> If ``B`` is a materialization of ``A``, we can say that ``B`` is a "more

Author: Hmm. The "we can say" construction is fairly common in presentations of formal definitions like this: it clearly indicates "heads up, here we are about to present and define a new term, with a precise meaning." I feel like the edit here doesn't communicate that as well.

> If ``B`` is a materialization of ``A``, we can say that ``B`` is a "more

Reviewer 2: I see your point. IMHO this whole document can only be understood as a series of assertions that define otherwise unknown concepts, however. In that context, I don't feel like the extra verbosity really adds anything.

> If ``B`` is a materialization of ``A``, we can say that ``B`` is a "more

Author: Fair enough, but in this case I prefer to leave it as is.

materialization.

A fully static type ``A`` is consistent with another fully static type ``B`` if
and only if they are the same type (``A`` is equivalent to ``B``.)

A gradual type ``A`` is consistent with a gradual type ``B``, and ``B`` is
consistent with ``A``, if and only if there exists some fully static type ``C``
which is a materialization of both ``A`` and ``B``.

``Any`` is consistent with every type, and every type is consistent with
``Any``. (This follows from the definitions of materialization and consistency
but is worth stating explicitly.)

The consistency relation is not transitive. ``tuple[int, int]`` is consistent
with ``tuple[Any, int]``, and ``tuple[Any, int]`` is consistent with
``tuple[str, int]``, but ``tuple[int, int]`` is not consistent with
``tuple[str, int]``.

The consistency relation is symmetric. If ``A`` is consistent with ``B``, ``B``
is also consistent with ``A``. It is also reflexive: ``A`` is always consistent
with ``A``.

The assignable-to (or consistent subtyping) relation
----------------------------------------------------

Given the materialization relation and the subtyping relation, we can define

> the **consistent subtype** relation over all types. A type ``B`` is a

Author: Not clear to me that this is an improvement. There is, after all, only one consistent subtype relation (within this framework) -- now we can define it.

consistent subtype of ``A``.

In the remainder of this specification, we will usually prefer the term
**assignable to** over "consistent subtype of". The two are synonymous, but
"assignable to" is shorter, and may communicate a clearer intuition to many
readers.

For example, ``Any`` is assignable to ``int``, because ``int`` is a
materialization of ``Any``, and ``int`` is a subtype of ``int``. The same
materialization also gives that ``int`` is assignable to ``Any``.

The assignable-to relation is not generally symmetric, though. If ``B`` is a
subtype of ``A``, then ``tuple[Any, B]`` is assignable to ``tuple[int, A]``,
because ``tuple[Any, B]`` can materialize to ``tuple[int, B]``, which is a
subtype of ``tuple[int, A]``. But ``tuple[int, A]`` is not assignable to
``tuple[Any, B]``.

Attributes and methods
----------------------

In Python, we can do more with objects at runtime than just assign them to
names, pass them to functions, or return them from functions. We can also
get/set attributes and call methods.

In the Python object model, the operations that can be performed on a value all
de-sugar to method calls. For example, ``a + b`` is syntactic sugar for either

> ``a.__add__(b)`` or ``b.__radd__(a)``.

Reviewer 2: We may as well be precise here, since this is a specification, after all. Special methods are always looked up on the class :P

```suggestion
``type(a).__add__(a, b)`` or ``type(b).__radd__(b, a)``.
```

Reviewer 3: This is also not pedantically correct, though:

```
class C(type):
    def __getattribute__(self, x):
        return lambda x, y: ("meta", x, y)

class A(metaclass=C):
    def __add__(self, other):
        return ("method", self, other)
    
    def __repr__(self) -> str:
        return "A()"

a = A()
b = A()
print(a + b)  # ('method', A(), A())
print(type(a).__add__(a, b))  # ('meta', A(), A())
```

Reviewer 2: I guess. Maybe we should just hedge slightly and say that it "approximately" desugars to ``a.__add__(b)`` or ``b.__radd__(a)``, and link to https://snarky.ca/unravelling-binary-arithmetic-operations-in-python/

Author: Ultimately I think it would be good for the typing spec to take a clear stand on whether overriding `__getattribute__` is or isn't part of the type system's (abstracted) model of the Python object model. Personally, I'm not sure it should be.

But this PR isn't the place to make that call. I'll adjust this wording to be accurate without getting too far into the weeds.

For a static type checker, accessing ``a.foo`` is a type error unless all
possible objects in the set represented by the type of ``a`` have the ``foo``
attribute. (We consider an implementation of ``__getattr__`` to be a getter for
all attribute names, and similarly for ``__setattr__`` and ``__delattr__``).

If all objects in the set represented by the fully static type ``A`` have a
``foo`` attribute, we can say that the type ``A`` has the ``foo`` attribute.

> If the type ``A`` of ``a`` in ``a.foo`` is a gradual type, it may not represent
> a single set of objects. In this case, ``a.foo`` is not a type error if and
> only if there exists a materialization of ``A`` which has the ``foo``
> attribute.

Reviewer 2: The order here feels backwards to me: you discuss the negation (the conditions required for a static type error not to be triggered) before discussing the conditions required for a static type error to be triggered:

```suggestion
If the type ``A`` of ``a`` in ``a.foo`` is a gradual type, it may not represent
a single set of objects. In this case, ``a.foo`` is a type error if and
only if there does not exist any materialization of ``A`` which has the
``foo`` attribute.
```

> Equivalently, we can say that ``a.foo`` is a type error unless the type of
> ``a`` is assignable to a type that has the ``foo`` attribute.

Reviewer 2: Why say that you _can_ say something when you could just... say it? ;)

```suggestion
Equivalently, ``a.foo`` is a type error unless the type of
``a`` is assignable to a type that has the ``foo`` attribute.
```

Author: Ok, in this case I agree, it adds nothing :) Removing.

every operation will be checked, but certain primitive operations in the
language such as attribute access or arithmetic are. Python is a
dynamically typed language.

**Gradual typing** is a way to combine static and dynamic typing.
Type-annotated Python allows opting in to static type checking at a fine level
of granularity, so that some type errors can be caught statically, without
running the program. Variables, parameters, and returns can optionally be given
a static type annotation. Even within the type of a single data structure,
static type checking is optional and granular. For example, a dictionary can be
annotated to enable static checking of the key type but only have dynamic
runtime checking of the value type.

A **gradual** type system is one in which a special "unknown" or "dynamic" type
is used to describe names or expressions whose types are not known statically.
In Python, this type is spelled :ref:`Any`. Because :ref:`Any` indicates a
statically unknown type, the static type checker can't check type correctness
of operations on expressions typed as :ref:`Any`. These operations are still
dynamically checked, via the Python runtime's usual dynamic checking.

> The Python type system also uses ``...`` within :ref:`Callable` types and

Reviewer 3: The glossary defines the term "gradual form" for these. It may be useful to also use that term here, so you don't have to repeat the references to Callable and tuple in the next section.

Author: Oh yes, thank you, I intended to do that after adding the glossary entry, but then forgot to get back to it.

subtype of ``B`` and ``B`` is a subtype of ``A``. This means that the set of
values represented by ``A`` is both a superset and a subset of the values
represented by ``B``, meaning ``A`` and ``B`` must represent the same set of
values.

We may describe a type ``B`` as "narrower" than a type ``A`` (or as a "proper
subtype" of ``A``) if ``B`` is a subtype of ``A`` and ``B`` is not equivalent
to ``A``. In the same scenario we can describe the type ``A`` as "wider" than
``B``, or a "proper supertype" of ``B``.

Materialization
---------------

Since :ref:`Any` represents an unknown static type, it does not represent any
known single set of values (it represents an unknown set of values). Thus it is
not in the domain of the subtype, supertype, or equivalence relations on static
types described above.

To relate gradual types more generally, we define a **materialization**
relation. Materialization transforms a "more dynamic" type to a "more static"
type. Given a gradual type ``A``, if we replace zero or more occurrences of
``Any`` in ``A`` with some gradual type (which can be different for each
occurrence of ``Any``), the resulting gradual type ``B`` is a materialization

> of ``A``. (We can also materialize a :ref:`Callable` type by replacing ``...``
> with a bracketed list of types, and materialize ``tuple[Any, ...]`` by

Reviewer 3: ```suggestion
of ``A``. (We can also materialize a :ref:`Callable` type by replacing it
with a callable types with a fully static parameter list, and materialize ``tuple[Any, ...]`` by
```
Nit: Materializing `Callable[..., Any]` doesn't necessarily mean replacing the `...` with a "bracketed list of types", because such a bracketed list can only express positional-only parameters. For example, given `f: Callable[..., int]`, a call `f(x=3)` is valid because the type of `f` could be materialized to a callable type that accepts an `x` keyword parameter.

materialization of ``Any``, and ``int`` is a subtype of ``int``. The same
materialization also shows that ``int`` is assignable to ``Any``.

The assignable-to relation is not generally symmetric, however. If ``B`` is a
subtype of ``A``, then ``tuple[Any, B]`` is assignable to ``tuple[int, A]``,
because ``tuple[Any, B]`` can materialize to ``tuple[int, B]``, which is a
subtype of ``tuple[int, A]``. But ``tuple[int, A]`` is not assignable to
``tuple[Any, B]``.

Attributes and methods
----------------------

In Python, we can do more with objects at runtime than just assign them to
names, pass them to functions, or return them from functions. We can also
get/set attributes and call methods.

In the Python object model, the operations that can be performed on a value all
desugar to method calls. For example, ``a + b`` is (roughly, eliding some
details) syntactic sugar for either ``type(a).__add__(a, b)`` or
``type(b).__radd__(b, a)``.

For a static type checker, accessing ``a.foo`` is a type error unless all
possible objects in the set represented by the type of ``a`` have the ``foo``
attribute. (We consider an implementation of ``__getattr__`` to be a getter for

> all attribute names, and similarly for ``__setattr__`` and ``__delattr__``).

Reviewer 3: ```suggestion
all attribute names, and similarly for ``__setattr__`` and ``__delattr__``.)
```
The whole sentence is within parentheses, so the period should also be.

Type system concepts
====================

> Static, dynamic, and gradual typing

Reviewer 4: First off, this section is excellent, and I learned a few things from reading it.

I have to admit that I got a little lost around "Materialization" on my first read-through because a lot of the terminology around gradual types was new to me. What helped me was to explicitly write out a table like this:

Fully static types | Gradual types
------------------------ | ---------------------
B is a subtype of A | B is assignable to A (or B is a consistent subtype of A)
B is a supertype of A | B is assignable from A (or B is a consistent supertype of A)
B is equivalent to A | B is consistent with A

I wonder if including something like this in the doc would also help other readers.

Author: I added this table. I put it after the description of all the relations, since it seemed necessary to present them before presenting the table. Let me know if this location would have been helpful in your initial reading, or if it should come sooner.

Reviewer 4: Yeah, I think that works well!

(respectively, parameter's type annotation or return type annotation).

We can say that a type ``B`` is "assignable to" a type ``A`` if ``B`` is a
consistent subtype of ``A``.

In the remainder of this specification, we will usually prefer the term
**assignable to** over "consistent subtype of". The two are synonymous, but
"assignable to" is shorter, and may communicate a clearer intuition to many
readers.

For example, ``Any`` is assignable to ``int``, because ``int`` is a
materialization of ``Any``, and ``int`` is a subtype of ``int``. The same
materialization also shows that ``int`` is assignable to ``Any``.

The assignable-to relation is not generally symmetric, however. If ``B`` is a
subtype of ``A``, then ``tuple[Any, B]`` is assignable to ``tuple[int, A]``,
because ``tuple[Any, B]`` can materialize to ``tuple[int, B]``, which is a
subtype of ``tuple[int, A]``. But ``tuple[int, A]`` is not assignable to
``tuple[Any, B]``.

Summary of type relations
-------------------------

The subtype, supertype, and equivalence relations establish a partial order on
fully static types. The analogous relations on gradual types (via

> materialization) are "assignable-to" (or consistent subtype), "assignable-from"
> (or consistent supertype), and "consistent with". We can visualize this analogy

Reviewer 2: ```suggestion
materialization) are "assignable-to" (or "consistent subtype"), "assignable-from"
(or "consistent supertype"), and "consistent with". We can visualize this analogy
```

component of a type. The detailed rules for these usages are discussed in their
respective sections of the specification. Collectively, along with :ref:`Any`,
these are :term:`gradual forms <gradual form>`.

This specification describes a gradual type system for Python.

Fully static and gradual types
------------------------------

We will refer to types that do not contain a :term:`gradual form` as a sub-part
as **fully static types**.

A **gradual type** can be a fully static type, :ref:`Any` itself, or a type
that contains a gradual form as a sub-part.

Fully static types
~~~~~~~~~~~~~~~~~~

A fully static type denotes a set of potential runtime values. For instance,
the fully static type ``object`` is the set of all Python objects. The fully
static type ``bool`` is the set of values ``{ True, False }``. The fully static
type ``str`` is the set of all Python strings; more precisely, the set of all
Python objects whose runtime type (``__class__`` attribute) is either ``str``
or a class that inherits directly or indirectly from ``str``. A :ref:`Protocol
<Protocols>` denotes the set of all objects which share a certain set of

> attributes and/or methods.

Author: So are Callables. I think this isn't the place to attempt a full enumeration; it's just a set of examples.

runtime values.

This may appear similar to the fully static type ``object``, which represents
the set of all Python objects, but it is quite different.

If an expression has the type ``object``, a static type checker should ensure
that operations on the expression are valid for all Python objects, or else
emit a static type error. This allows very few operations! For example, if
``x`` is typed as ``object``, ``x.foo`` should be a static type error because
not all Python objects have an attribute ``foo``.

An expression typed as :ref:`Any`, on the other hand, should be assumed to have
_some_ specific static type, but _which_ static type is not known. A static
type checker should not emit static type errors on an expression or statement
if :ref:`Any` might represent a static type which would avoid the error. (This
is defined more precisely below, in terms of materialization and
assignability.)

Similarly, a gradual type such as ``tuple[int, Any]`` (see :ref:`tuples`) or
``int | Any`` (see :ref:`union-types`) does not represent a single set of
Python objects; rather, it represents a (bounded) range of possible sets of
values.

> In the same way that ``Any`` does not represent "the set of all Python objects"

Reviewer 2: ```suggestion
In the same way that :ref:`Any` does not represent "the set of all Python objects"
```

but rather "an unknown set of objects", ``tuple[int, Any]`` does not represent

> "the set of all length-two tuples whose first element is an integer." That is a

Reviewer 3: Note this is a British/American difference; see https://en.wikipedia.org/wiki/Quotation_marks_in_English#Logical_quotation.

> "the set of all length-two tuples whose first element is an integer." That is a

Reviewer 2: Bah, well, feel free to do the ~~wrong~~ American thing if you prefer then :P

> "the set of all length-two tuples whose first element is an integer." That is a

Reviewer 3: Though as the Wikipedia article notes, the "British" style also gets widely used in American sources, especially in a computing context. Personally I prefer to use this style too.

> "the set of all length-two tuples whose first element is an integer." That is a

Author: The British approach makes logical sense to me, but looks wrong because it's now how I was taught to write ;) I'll try to retrain my instincts.

dictionary keys and values. These are called "structural types". A structural
type may be a subtype of another type without any inheritance or subclassing
relationship, simply because it meets all the requirements of the supertype,
and perhaps adds more, thus representing a subset of the possible values of the
supertype. This is "structural subtyping".

Although the means of specifying the set of values represented by the types
differs, the fundamental concepts are the same for both nominal and structural
types: a type represents a set of possible values and a subtype represents a
subset of those values.

Materialization
---------------

Since :ref:`Any` represents an unknown static type, it does not represent any
known single set of values (it represents an unknown set of values). Thus it is
not in the domain of the subtype, supertype, or equivalence relations on static
types described above.

To relate gradual types more generally, we define a **materialization**
relation. Materialization transforms a "more dynamic" type to a "more static"
type. Given a gradual type ``A``, if we replace zero or more occurrences of

> ``Any`` in ``A`` with some gradual type (which can be different for each

Author: "(possibly gradual)" wouldn't be right since all types are gradual, but I still do think this is clearer without the word gradual there 👍

occurrence of ``Any``), the resulting gradual type ``B`` is a materialization
of ``A``. (We can also materialize a :ref:`Callable` type by replacing ``...``
with any type signature, and materialize ``tuple[Any, ...]`` by replacing it

> with a determinate-length tuple type.)

Reviewer 3: Or by an indeterminate-length tuple type such as `tuple[int, ...]`, which is also a fully static type.

Reviewer 3: Thanks, I see what you mean now; this is fine.

``Any`` is consistent with every type, and every type is consistent with
``Any``. (This follows from the definitions of materialization and consistency
but is worth stating explicitly.)

The consistency relation is not transitive. ``tuple[int, int]`` is consistent
with ``tuple[Any, int]``, and ``tuple[Any, int]`` is consistent with
``tuple[str, int]``, but ``tuple[int, int]`` is not consistent with
``tuple[str, int]``.

The consistency relation is symmetric. If ``A`` is consistent with ``B``, ``B``
is also consistent with ``A``. It is also reflexive: ``A`` is always consistent
with ``A``.

.. _`assignable`:

The assignable-to (or consistent subtyping) relation
----------------------------------------------------

Given the materialization relation and the subtyping relation, we can define
the **consistent subtype** relation over all types. A type ``B`` is a
consistent subtype of a type ``A`` if there exists a materialization ``A'`` of
``A`` and a materialization ``B'`` of ``B``, where ``A'`` and ``B'`` are both
fully static types, and ``B'`` is a subtype of ``A'``.

Consistent subtyping defines "assignability" for Python.  An expression can be

> assigned to a variable (including passed as a parameter or returned from a

Reviewer: Nit: "passed as a parameter" should probably be "passed as an argument" or "passed to a parameter"

     - ``A`` is assignable from (or a consistent supertype of) ``B``
   * - ``B`` is :term:`equivalent` to ``A``
     - ``B`` is :term:`consistent` with ``A``

We can also define equivalence on gradual types. Two gradual types ``A`` and
``B`` are equivalent (that is, the same gradual type, not merely consistent
with one another) if and only if all materializations of ``A`` are also
materializations of ``B``, and all materializations of ``B`` are also
materializations of ``A``.

Attributes and methods
----------------------

In Python, we can do more with objects at runtime than just assign them to
names, pass them to functions, or return them from functions. We can also
get/set attributes and call methods.

In the Python object model, the operations that can be performed on a value all
desugar to method calls. For example, ``a + b`` is (roughly, eliding some
details) syntactic sugar for either ``type(a).__add__(a, b)`` or
``type(b).__radd__(b, a)``.

For a static type checker, accessing ``a.foo`` is a type error unless all
possible objects in the set represented by the type of ``a`` have the ``foo``
attribute. (We consider an implementation of ``__getattr__`` to be a getter for

> all attribute names, and similarly for ``__setattr__`` and ``__delattr__``.)

Author: I agree this section is slightly awkward currently, because I don't actually want to give a full specification of attribute access on types (that's deserving of its own chapter, for sure in a separate PR), but I do think that it's important to understanding the meaning of types (which is the purpose of the concepts doc) to have _some_ discussion of attribute access, not just assignability, since this is such a common operation on types.

So I definitely don't want to get into `__getattribute__` here; if anything I'd prefer to not even digress into `__getattr__`, because it does start to sound more like a specification of attribute access, which I don't want to get into here.

I think my preferred option would be to replace this parenthetical with just a note that attribute access is actually more complicated than that, and a link to the data model. I'll push that change; hopefully that's agreeable to everyone who has already done their final review.

> all attribute names, and similarly for ``__setattr__`` and ``__delattr__``.)

Author: I ended up leaving in the mention of `__getattr__` and `__setattr__`, since it has already been reviewed and seems more useful to have it than not.

I added a link to the data model doc and an acknowledgment that there are more complexities.

In the future if we do have a full spec of attribute access, this section could perhaps be slightly shortened and link to that instead.

type system. The type system assigns a type to all expressions in the language
and verifies that their uses obey the typing rules. Normally, a program that is
not well-typed (i.e., one that contains a type error) will not run. Java and
C++ are examples of statically-typed object-oriented languages.

A **dynamically-typed** programming language does not run a type-checker before
running a program. Instead, it will check the types of values before performing
operations on them at runtime. This is not to say that the language is
"untyped". Values at runtime have a type and their uses obey typing rules. Not
every operation will be checked, but certain primitive operations in the
language such as attribute access or arithmetic will be.  Python was
historically a dynamically-typed language.

**Gradual typing** is the name for a specific way to combine static and dynamic
typing.  It allows mixing static and dynamic checking at a fine level of
granularity.  For instance, in Python, individual variables, parameters, and
return values can be either statically or dynamically typed.  Data structures
such as objects can have a mix of static and dynamic checking.  As an example,

> a Python dictionary could choose to have static checking of the key type but

Reviewer 3: It's odd to me to describe this as a choice by the dictionary itself, rather than by the user. I'm not sure I agree with the earlier sentence "in Python, individual variables, parameters, and return values can be either statically or dynamically typed". In Python the language all values are dynamically typed; we're describing a static type system, but that system is separate from the language itself.

Reviewer 5: This is a place where I'd prefer to be a little more careful about using the word "value" - I don't like saying that values have types.

I know in this case "return value" is being used as a synonym of "returned expression" and "dictionary" is being used as a synonym for "dictionary term", but I've seen discussion from users who haven't really internalized that static types exist on terms of code, not on the values flowing through the runtime themselves.

As a result I'd like us to to be pretty careful about our language.

Reviewer 6: @stroxler in a sense values do have types.  If static types denote sets of values, then a value has the type of the static type denotations it is a member of.

We more or less need that idea if we want to talk about "soundness".  The type system is sound if, whenever we have an expression `e` of type `T`, then the evaluation of `e` will produce a value that agrees with `T` in some sense.

Author: I've attempted a re-wording of this paragraph, taking into account the concerns mentioned here. Here's my effort:

> **Gradual typing** is the name for a specific way to combine static and dynamic
typing. It allows mixing static and dynamic checking at a fine level of
granularity. For instance, in type-annotated Python, individual variables,
parameters, and returns can optionally be given a static type. Data structures
such as objects can have a mix of static and dynamic-only checking. As an
example, a Python dictionary could be annotated to have static checking of the
key type but only dynamic checking of the value type.

dynamic checking of the value type.

In gradual typing, a dynamically typed value is indicated by a special
"unknown" or "dynamic" type.  In Python, the unknown type is spelled
:ref:`Any`. It indicates to the static type checker that this value should not
be subject to static checking.  The system should not signal a static type
error for use of an expression with type :ref:`Any`.  Instead, the expression's
value will be dynamically checked, according to the Python runtime's usual
checks on the types of runtime values.

This specification describes a gradual type system for Python.

Static, dynamic, and gradual types
----------------------------------

> We will refer to types that do not contain :ref:`Any` as a sub-part as **static

Author: Yeah, I tried something different here to avoid the awkwardness of "fully static type," but I agree that it didn't work out that well, and isn't consistent with the literature. I like your suggestion, I'll try a re-wording to use the terms in that way.

False }``. The Python static type ``str`` is the set of all Python strings;
more precisely, the set of all Python objects whose runtime type (i.e.
``__class__`` attribute) is either ``str`` or a class that inherits ``str``,
including transitively; i.e. a type with ``str`` in its method resolution
order. Static types can also be specified in other ways. For example,
:ref:`Protocols` specify a static type which is the set of all objects which
share a certain set of attributes and/or methods.

The dynamic type
~~~~~~~~~~~~~~~~

The dynamic type :ref:`Any` represents an unknown static type. It denotes some
unknown set of values.

At first glance, this may appear similar to the static type ``object``, which
represents the set of all Python objects. But it is quite different.

If a value has the static type ``object``, a static type checker should ensure
that all operations on the value are valid for all Python objects, or else emit
a static type error. This allows very few operations! For example, if ``x`` is
typed as ``object``, ``x.foo`` should be a static type error, because not all
Python objects have an attribute ``foo``.

> A value typed as :ref:`Any`, on the other hand, should be assumed to have

Author: I would say that the concept of "static type", as defined here, always includes any distinct set of Python objects, regardless of whether that set is expressible in the type annotation syntax.

As a concrete example, I think that the type `A & B` (the intersection of A and B) is already today a static type in the Python type system, even though today this type is not (yet) expressible in an annotation expression.

_some_ specific static type, but _which_ static type is not known. A static
checker should not emit any errors that depend on assuming a particular static
type; a static checker should instead assume that the runtime is responsible
for checking the type of operations on this value, as in a dynamically-typed
language.

The subtype relation
--------------------

A static type ``B`` is a **subtype** of another static type ``A`` if (and only
if) the set of values represented by ``B`` is a subset of the set of values
represented by ``A``. Because the subset relation on sets is transitive and
reflexive, the subtype relation is also transitive (if ``A`` is a subtype of
``B`` and ``B`` is a subtype of ``C``, then ``A`` is a subtype of ``C``) and

> reflexive (``A`` is always a subtype of ``A``).

Reviewer 6: There's an annoying caveat about `int` and `float` in Python.  They have a subtype relation by rule as an exception to semantic subtyping.  (Or else we repair that by saying that the set of values denoted by `float` includes all the `int` values.)

I don't know if you can do footnotes or end notes in this doc.  It seems kind of intrusive to mention this special technical case here.

Author: Yes, I would rather handle this case in a later section of the spec specific to ints and floats, I don't think it needs to be mentioned here. (It may already be described somewhere in the spec, I haven't yet looked.) We may need to have some discussion specifically about the intended meaning of this special case. (For instance, one possible interpretation is that the annotation `float` is in fact understood to mean `float | int`, as a special case in the annotation syntax.)

Reviewer 7: There are also other ways to handle literal types which do not require `int <: float`.

The **supertype** relation is the inverse of subtype: ``A`` is a supertype of

> ``B`` if (and only if) ``B`` is a subtype of ``A``; or equivalently, if (and

Reviewer 6: I prefer "if and only if" over "if (and only if)".

In any case, be consistent.

only if) the set of values represented by ``A`` is a superset of the values
represented by ``B``. The supertype relation is also transitive and reflexive.

We also define an **equivalence** relation on static types: the types ``A`` and
``B`` are equivalent (or "the same type") if (and only if) ``A`` is a subtype
of ``B`` and ``B`` is a subtype of ``A``. This means that the set of values
represented by ``A`` is both a superset and a subset of the values represented
by ``B``, meaning ``A`` and ``B`` must represent the same set of values.

We may describe a type ``B`` as "narrower" than a type ``A`` if ``B`` is a
subtype of ``A`` and ``B`` is not equivalent to ``A``.

The consistency relation
------------------------

Since :ref:`Any` (the dynamic type) represents an unknown static type, it does
not represent any known single set of values, and thus it is not in the domain
of the subtype, supertype, or equivalence relations on static types described
above.

> We define a **materialization** relation from gradual types to gradual and

Reviewer 6: Let's give an intuitive explanation of what materialization is: if B is a materialization of A then B is a "more static" version of A, A is a "more dynamic" version of B.

Also, let's use a convention that letters closer to the beginning of the alphabet are either generally more static or generally more dynamic (it helps the reader to keep the directions of things straight, especially with things like contravariance).

I think it's important to highlight that different occurrences of Any can be replaced with different types.  We could do that by adding "... replacing ... with some gradual or static type, possibly different for each occurrence, results in...".

Another possibility is to define a single step of replacement and make materialization the transitive closure of that relation.  Then you can get from `tuple[Any, Any]` to `tuple[int, str]` in two steps of replacement.  I'm not sure if that's better.

Also, this relation is almost reflexive (and maybe we want it to be actually reflexive).  The definition written here allows replacing one occurrence of `Any` with `Any` so gradual types are materializations of themselves.  But this definition doesn't make static types materializations of themselves.  That could be repaired by saying "zero or more" and making this a relation on all types, or with the single step of replacement and saying materialization is the reflexive transitive closure.

Author: I think I'm happy enough with fixing this with your "zero or more" suggestion, but open to explicitly adding the single-step relation if it looks preferable to you after seeing my update.

static types as follows: if replacing one or more occurrences of ``Any`` in
gradual type ``A`` with some gradual or static type results in the gradual or
static type ``B``, then ``B`` is a materialization of ``A``. For instance,
``tuple[int, str]`` (a static type) and ``tuple[Any, str]`` (a gradual type)
are both materializations of ``tuple[Any, Any]``. If ``B`` is a materialization
of ``A``, we can say that ``B`` is a more precise type than ``A``.

We define a **consistency** relation over all types (the dynamic type, gradual
types, and static types.)

> A static type ``A`` is consistent with another static type ``B`` only if they

Reviewer 6: This is an "if and only if".

are the same type (``A`` is equivalent to ``B``.)

The dynamic type ``Any`` is consistent with every type, and every type is
consistent with ``Any``.

> A gradual type ``C`` is consistent with a gradual or static type ``D``, and

Reviewer 6: Why C and D, and not A and B?

Author: No good reason. I switched it to A and B.

``D`` is consistent with ``C``, if ``D`` is a materialization of ``C``.

> The consistency relation is not transitive. ``tuple[int, int]`` and

Reviewer 6: This isn't stated quite correctly.  Transitivity would have: `tuple[int, int]` is consistent with `tuple[Any, int]` and `tuple[Any, int]` is consistent with `tuple[str, int]` but `tuple[int, int]` is not consistent with `tuple[str, int]`.

Because the relation is symmetric, it amounts to the same thing but we haven't said it's symmetric yet. And relying on symmetry hides the failure of transitivity a little bit.  (I think it's important to realize that transitivity breaks because you can go "up and back down" the materialization relation.)

> ``tuple[str, int]`` are both consistent with ``tuple[Any, int]]``, but

Reviewer 3: ```suggestion
``tuple[str, int]`` are both consistent with ``tuple[Any, int]``, but
```

``tuple[str, int]`` is not consistent with ``tuple[int, int]``.

The consistency relation is symmetric. If ``A`` is consistent with ``B``, ``B``
is also consistent with ``A``. It is also reflexive: ``A`` is always consistent
with ``A``.

The consistent subtype relation
-------------------------------

> Given the consistency relation and the subtyping relation, we define the

Reviewer 6: This doesn't use consistency, it uses materialization so you could say "Given the materialization relation and the subtyping relation".

Do we actually need consistency for anything?  We might be able to get away with leaving it out.

Author: Good point.

I'd rather not remove consistency just yet, until I've had a chance to go over the rest of the spec and normalize the terms used, so I'm more confident we don't need it. I suspect there may be some cases (e.g. related to invariance, or typevar constraints) where we do need it? But I'm happy to remove it if it turns out we don't need it.

Reviewer 6: Oh yeah, TypeVar constraints was the one place that I thought we needed it.  I also see that I didn't really get consistent subtyping right for invariant or contravariant generics in my doc.

preorder on gradual types.

We also define a **consistency** relation on gradual types.

A static type ``A`` is consistent with another static type ``B`` if and only if
they are the same type (``A`` is equivalent to ``B``.)

A gradual type ``A`` is consistent with a gradual type ``B``, and ``B`` is
consistent with ``A``, if and only if ``B`` is a materialization of ``A`` or
``A`` is a materialization of ``B``.

The dynamic type ``Any`` is consistent with every type, and every type is
consistent with ``Any``. (This must follow from the above definitions of
materialization and consistency, but is worth stating explicitly.)

The consistency relation is not transitive. ``tuple[int, int]`` is consistent
with ``tuple[Any, int]`` and ``tuple[Any, int]`` is consistent with
``tuple[str, int]``, but ``tuple[int, int]`` is not consistent with
``tuple[str, int]``.

The consistency relation is symmetric. If ``A`` is consistent with ``B``, ``B``
is also consistent with ``A``. It is also reflexive: ``A`` is always consistent
with ``A``.

The consistent subtype relation
-------------------------------

Given the materialization relation and the subtyping relation, we define the

> **consistent subtype** relation over all types. A type ``A`` is a consistent

Reviewer 3: A thought: "Consistent subtyping" is a central concept in the type system but the name is awkward. There are more types that are a consistent subtype of a type A than there are types that are subtypes of A, which is contrary to what I'd expect. (Someone not familiar with the theory might think that some subtypes are "consistent subtypes" and some are "inconsistent subtypes".)

What do you think of naming this relation "assignability" or even "compatibility" instead? That would be more intuitive in the rest of the spec.

Author: I like this. I do think it will result in more natural (and clearer to most readers not familiar with gradual typing literature) usage throughout the rest of the spec. I would like to mention the term "consistent subtyping" here (for the benefit of those familiar with the literature, but it doesn't need to be the preferred term in the spec.

I'll try a reframe of the last couple sections along these lines.

Author: I think I prefer "assignable to" over "compatible with." The latter (to me) suggests a symmetric relation, which consistent subtyping is not: A can be assignable to B when B is not assignable to A. This would make me tend to think "compatible with" is more like "consistent with" than "consistent subtype of." So my leaning is to avoid the term "compatible" and instead prefer "assignable to." What do you think?

Reviewer 3: Agree "assignable" is better.

Reviewer 6: Consistent subtyping is the awkward term from the literature.  (For historical reasons: there was first "consistency" for languages that didn't have subtyping.  Then when languages with subtyping were considered, they used "consistent subtyping" for the combination of the two.)

I do like "assignability" (as long as we're sure that it really matches what can be assigned). It's the term I use with people when I explain why we care about consistent subtyping.

By the way, I do like this more "modern" factoring of assignability into subtyping and materialization.  We can make it into a motto: assignability is subtyping plus materialization.

You can assign a value of a subtype, and you can assign a value of a more or less dynamic type.

Type system concepts
====================

Static, dynamic, and gradual typing
-----------------------------------

A **statically-typed** programming language runs a type-checker before running
a program. The program is required to be well-typed according to the language's
type system. The type system assigns a type to all expressions in the language
and verifies that their uses obey the typing rules. Normally, a program that is
not well-typed (i.e., one that contains a type error) will not run. Java and
C++ are examples of statically-typed object-oriented languages.

A **dynamically-typed** programming language does not run a type-checker before
running a program. Instead, it will check the types of values before performing
operations on them at runtime. This is not to say that the language is
"untyped". Values at runtime have a type and their uses obey typing rules. Not
every operation will be checked, but certain primitive operations in the
language such as attribute access or arithmetic will be. Python was
historically a dynamically-typed language.

**Gradual typing** is the name for a specific way to combine static and dynamic

> typing. It allows mixing static and dynamic checking at a fine level of

Reviewer 8: It can be a coarse granularity too. We have macro gradual typing which enables us to type at the module level (such as Typed Racket) and micro-gradual typing which allows us to type at the expression level (such as Reticulated Python).

Author: Good point. Do you think this is important to clarify here? I guess it is still true that gradual typing _allows_ mixing at fine granularity.

Reviewer 8: Oh that's true! 
If we're not talking about typed/untyped boundaries, runtime type checks or performance, then it may not be important to mention.

Reviewer 6: We shouldn't say things that are wrong, and I think it's a good point that this isn't quite right.  Probably just better to say "Python allows" instead of "It allows" and they we don't need to mention Python again in the next sentence.

granularity. For instance, in type-annotated Python, individual variables,
parameters, and returns can optionally be given a static type. Data structures

> such as objects can have a mix of static and dynamic-only checking. As an

Reviewer 8: Are we discussing two different concepts here? The concept that a given annotation can contain the "Dyn" type, but also the granularity required by the type-checker for mixing typed/untyped code?

Author: I'm not sure I understand the distinction between what "a given annotation can contain" vs "the granularity required by the type checker." Can you say more?

I see this section as a fairly informal introduction to the concept that Any and static types can coexist at fine granularity.

Reviewer 8: I was more wondering what we are talking about from L26 to L32. Are we describing partial type annotations such as `Dyn -> Int` or are we still talking about granularity? also is there a distinction between variables and parameters vs. objects that we are trying to highlight? overall, I am just finding those lines hard to grasp

example, a Python dictionary could be annotated to have static checking of the
key type but only dynamic checking of the value type.

> In gradual typing, a dynamically typed value is indicated by a special

Reviewer 8: Since here, we are are describing the behavior of the `dyn` type with respect to how it gets type checked, could we say something about the design philosophy behind the typechecker behavior? 

For example, the reason expressions that are annotated with `dyn` are not be subject to static type-checking, is that we optimistically assume that that the value for those expressions will have appropriate types at runtime.

Author: Right, we could also discuss the gradual guarantee as a way to motivate this behavior. I guess essentially both framings of the idea boil down to the assumption that we are adding static types to an already-working untyped program.

Reviewer 6: I'd be careful not to frame this misleadingly.  It kind of sounds like the programmer should use `Any` for code that they believe is correct.

A lot of gradual typing systems will insert runtime checked casts of dynamic values, explicitly not assuming that they are correct.

There are all kinds of reasons that Python programmers might write new code with `Any`.

Author: I added a short paragraph on the gradual guarantee. I'm not sure we need it, but it's probably useful to mention.

``B`` and ``B`` is a subtype of ``A``, then ``C`` is a subtype of ``A``) and
reflexive (``A`` is always a subtype of ``A``).

The **supertype** relation is the inverse of subtype: ``A`` is a supertype of
``B`` if and only if ``B`` is a subtype of ``A``; or equivalently, if and only
if the set of values represented by ``A`` is a superset of the values
represented by ``B``. The supertype relation is also transitive and reflexive.

We also define an **equivalence** relation on static types: the types ``A`` and
``B`` are equivalent (or "the same type") if and only if ``A`` is a subtype
of ``B`` and ``B`` is a subtype of ``A``. This means that the set of values
represented by ``A`` is both a superset and a subset of the values represented
by ``B``, meaning ``A`` and ``B`` must represent the same set of values.

We may describe a type ``B`` as "narrower" than a type ``A`` (or as a "strict
subtype" of ``A``) if ``B`` is a subtype of ``A`` and ``B`` is not equivalent
to ``A``.

The consistency relation
------------------------

Since :ref:`Any` represents an unknown static type, it does not represent any
known single set of values, and thus it is not in the domain of the subtype,

> supertype, or equivalence relations on static types described above.

Reviewer 7: We probably have to something about how `Any` participates in subtyping, though: is `Any` a subtype of `Any`?

A practical example where this pops up is union normalization, where we want `T | Any | Any` be equivalent to `T | Any`. With the current definition, these two unions are distinct, I think.

Author: Yeah, great point, will add that.

> We define a **materialization** relation on gradual types as follows: if

Author: Yes, good idea, I'll work to rearrange this a bit to be a gentler introduction to the idea.

replacing zero or more occurrences of ``Any`` in gradual type ``A`` with some
gradual type (which can be different for each occurrence of ``Any``) results in
the gradual type ``B``, then ``B`` is a materialization of ``A``. For instance,
``tuple[int, str]`` (a static type) and ``tuple[Any, str]`` (a gradual type)
are both materializations of ``tuple[Any, Any]``. ``tuple[int, str]`` is also a
materialization of ``tuple[Any, str]``.

If ``B`` is a materialization of ``A``, we can say that ``B`` is a "more
static" type than ``A``, and ``A`` is a "more dynamic" type than ``B``.

The materialization relation is both transitive and reflexive, so it defines a
preorder on gradual types.

We also define a **consistency** relation on gradual types.

A static type ``A`` is consistent with another static type ``B`` if and only if
they are the same type (``A`` is equivalent to ``B``.)

A gradual type ``A`` is consistent with a gradual type ``B``, and ``B`` is
consistent with ``A``, if and only if ``B`` is a materialization of ``A`` or

> ``A`` is a materialization of ``B``.

Reviewer 8: Hmmm, I'm confused here.  `Int -> Dyn` is not related by precision to `Dyn -> Int`, yet they are consistent. Am I missing something? I think of the precision relation and the materialization relation as the same thing.

Author: Ah, thank you! I think this is in fact worded too restrictively. Consistency should rather require that there exists a common static type `C` which is a materialization of both `A` and `B`. Does that seem right, or is there a better way to specify it? I'll check back to see what Lanvin says exactly here.

Reviewer 6: Yeah, this doesn't seem quite right.  I think we need a third type `C` which is a materialization of both `A` and `B`, similar to what we have for consistent subtyping.

The dynamic type ``Any`` is consistent with every type, and every type is
consistent with ``Any``. (This must follow from the above definitions of
materialization and consistency, but is worth stating explicitly.)

The consistency relation is not transitive. ``tuple[int, int]`` is consistent
with ``tuple[Any, int]`` and ``tuple[Any, int]`` is consistent with
``tuple[str, int]``, but ``tuple[int, int]`` is not consistent with
``tuple[str, int]``.

> The consistency relation is symmetric. If ``A`` is consistent with ``B``, ``B``

Reviewer 8: Will we also mention that those concepts extend to programs? for example, if we want to talk about the gradual guarantee, then we want to talk about program A being more static than program B.

Author: I referenced this briefly in the section I added about the gradual guarantee.

is also consistent with ``A``. It is also reflexive: ``A`` is always consistent
with ``A``.

The consistent subtype relation
-------------------------------

Given the materialization relation and the subtyping relation, we define the
**consistent subtype** relation over all types. A type ``A`` is a consistent
subtype of a type ``B`` if there exists a materialization ``A'`` of ``A`` and a
materialization ``B'`` of ``B``, where ``A'`` and ``B'`` are both static types,
and ``A'`` is a subtype of ``B'``.

> For example, ``Any`` is a consistent subtype of ``int``, because ``int`` is a

Reviewer 8: could we give an example to highlight the asymmetry of this relation? because it seems this example highlights the "consistency" part but not the "subtyping" part?

Author: Great point. I'm finding the examples a bit tricky since we haven't really introduced the syntactic type forms yet at this point. But I've already assumed an understanding of the tuple type in previous examples, so I can use that here as well and show an asymmetric example here.

Type system concepts
====================

Static, dynamic, and gradual typing
-----------------------------------

A **statically-typed** programming language runs a type-checker before running
a program. The program is required to be well-typed according to the language's
type system. The type system assigns a type to all expressions in the language
and verifies that their uses obey the typing rules. Normally, a program that is
not well-typed (i.e., one that contains a type error) will not run. Java and
C++ are examples of statically-typed object-oriented languages.

A **dynamically-typed** programming language does not run a type-checker before

> running a program. Instead, it will check the types of values before performing

Author: Fixing the cases of this that I notice; if there are other cases that stand out to you that I miss, feel free to point them out.

operations on them at runtime. This is not to say that the language is
"untyped". Values at runtime have a type and their uses obey typing rules. Not
every operation will be checked, but certain primitive operations in the
language such as attribute access or arithmetic will be. Python was
historically a dynamically-typed language.

**Gradual typing** is the name for a specific way to combine static and dynamic

> typing. Type-annotated Python allows mixing static and dynamic checking at a

Author: I'm not sure I see why "mix" is confusing. Despite the fact that dynamic type checking always occurs, once we also add some static type checking, there is a mix. Some parts of the program are now (also) statically checked, others rely solely on the dynamic checking.

But I will try to re-word to avoid the phrasing.

fine level of granularity. Individual variables, parameters, and returns can
optionally be given a static type. Data structures such as objects can have a
mix of static and dynamic-only checking. For example, a dictionary could be
annotated to have static checking of the key type but only dynamic checking of
the value type.

> In gradual typing, a dynamically typed term is indicated by a special "unknown"

Author: The word "term" is often used in programming languages literature, more or less synonomously with "expression." See e.g. https://cs.stackexchange.com/a/64702

If the word doesn't communicate well, I'll find alternatives.

From the perspective of a static type checker, I think it makes sense to refer to things typed as Any as "dynamically checked," because the static type checker generally isn't able to check them, and thus the program is relying solely on dynamic checking of those values. (The fact that all values are dynamically checked isn't really relevant from the perspective of the static checker.) Perhaps the clarification that this is "from the perspective of the static checker" is missing. I'll see how I can rephrase.

> In gradual typing, a dynamically typed term is indicated by a special "unknown"

Author: I've avoided the word "term", and adjusted quite a bit of wording. Curious if you like the new version better.

> or "dynamic" type.  In Python, the unknown type is spelled :ref:`Any`. It

Author: My goal here is to get the basic concepts established without getting into too many details of particular parts of the type system. If we can establish a clear understanding of `Any`, then the tuple and callable docs can rely on that to discuss their use of `...` by analogy.

That said, I think it makes sense to briefly mention here these additional gradual typing constructs, with links to the appropriate detailed documentation.

> or "dynamic" type.  In Python, the unknown type is spelled :ref:`Any`. It

Reviewer: That makes sense, but that means that it's important that the basic concept should extend to other examples. I contend that the current definition of `Any` ("should not be subject to static checking") doesn't extend to these other cases well. That's a bit of a red flag for me, and it's why I'm suggesting that we may need a better way of defining `Any`.

> indicates to the static type checker that this term should not be subject to

Author: Really appreciate the concern for precise language!

This description of Any is intended to be an introductory intuition. The more precise definition of `Any` in this spec is intended to be the later one: that Any represents an unknown static type. And even more precision is offered via the materialization and assignability relations, which define exactly what it means to say "the set of all values that will satisfy the typing rules."

I think one place where more clarity would be beneficial is in the description of where assignability applies. We currently mention assignment, function calls, and returns, but we don't yet discuss materialization or assignability in contexts like method or attribute lookup.

That said, I think I agree that "should not be subject to static type checking" is too imprecise here. I'll give some thought to clearer language.

static checking.  The system should not signal a static type error for use of a
term with type :ref:`Any`.  Instead, the term's type will be dynamically
checked, according to the Python runtime's usual checks on the types of runtime
values.

This specification describes a gradual type system for Python.

Static, dynamic, and gradual types
----------------------------------

> We will refer to types that do not contain :ref:`Any` as a sub-part as **static

Author: According to the definitions I currently give in this PR, only `Any` is "the dynamic type." All types (including `Any`, `list[Any]`, and e.g. `int`) are "gradual types." Only types that don't involve `Any` are "static types." (The category "gradual type" encompasses the category "static type.")

It is important to make this distinction somehow because it is fundamental to the definitions of subtyping, materialization, and consistency. Only static types represent a specific set of Python objects and behave according to a set-theoretic view of subtyping. Assignability (or consistent subtyping) requires both materialization of gradual types to static types and the subtype relation on static types.

Using the term "dynamic type" for Any is very well established in the literature (in many treatments it is actually called `dynamic` or `Dyn`). I think it's also intuitive because Any does defer type checking to the runtime, which is precisely what "dynamic typing" or "dynamically typed" means. So I don't see a good case for avoiding the use of "dynamic type" to refer to Any.

If Any is "the dynamic type" it seems logical to name types not involving Any as "static types." This is also reasonable because these are the only types that appear in fully static (i.e. non-gradual) type systems. Kevin's doc used the term "fully static types" for these; I'd be fine with that term if it seems clearer. Also open to considering other options if there are suggestions.

> We will refer to types that do not contain :ref:`Any` as a sub-part as **static

Reviewer: If `Any` is the only example of "dynamic type" then why does it need a separate term? Let's just call it `Any` then. 

TypeScript has an `any` type that is analogous to the one in Python's type system, but I don't think you'll find any reference to "dynamic type" in TypeScript's specification.

I'm good with the term "gradual type" to mean "any type that incorporates `Any`, `...` or similar gradual typing affordances. 

We could then eliminate the term "static type", which I think is really confusing given that these are all "static type annotations". The term "non-gradual type" could be used in the few places where the spec needs to make a clear distinction between gradual and non-gradual types.

> We will refer to types that do not contain :ref:`Any` as a sub-part as **static

Author: I find it hard to write this spec without a term for "types that don't contain Any", because that's a really important concept -- only these types represent a single set of objects, and participate directly in subtyping. I really don't like the way it reads to only use a negative description (e.g. "non-gradual") of such types. Also, it reads much better to define "gradual type" as including all types, otherwise we end up with a lot of "gradual and non-gradual types" all over the place. (Could maybe just be "types" but I think this is not clear enough.)

I switched from "static type" to "fully static type" in the hopes that this might help address your concern that "static type" is confusing.

I still think that it is highly valuable to describe the materialization preorder as a continuum from "more dynamic" types (which contain more non-statically-known components) to "more static" types (which contain fewer), all the way to "fully static" types (which contain no non-statically-known components.) And I still feel that the terms "dynamic" and "static" are quite sensible in this context, as the latter are more "statically known" and "statically checked", while the former are less statically known, and thus (to some degree) more reliant on the dynamic checking which applies to all values at runtime.

Will be grateful for your review of the latest version.

types**.

> :ref:`Any` itself is the **dynamic type**.

Author: I don't think the fact that all types are checked at runtime in Python is a reason to avoid the term "dynamic type" for Any. Any expresses that we are _relying_ on that "dynamic typing" behavior here, rather than relying on static type checking.

I don't think the use of the term "dynamic" for Any is related to the insertion of runtime checked casts in sound gradually typed languages. The naming is more fundamental to the mix of static and dynamic typing which characterizes gradual type systems, whether sound (inserting runtime checked casts) or not.

A **gradual type** can be a static type, :ref:`Any` itself, or a type that
contains :ref:`Any` as a sub-part.

Static types
~~~~~~~~~~~~

A static type can intuitively be understood as denoting a set of runtime
values. For instance, the Python static type ``object`` is the set of all
Python objects. The Python static type ``bool`` is the set of values ``{ True,
False }``. The Python static type ``str`` is the set of all Python strings;
more precisely, the set of all Python objects whose runtime type (i.e.
``__class__`` attribute) is either ``str`` or a class that inherits ``str``,
including transitively; i.e. a type with ``str`` in its method resolution
order.

> Static types can also be specified in other ways. For example, :ref:`Protocols`

Author: In a sense the discussion of `str` implicitly presents "a way to specify a static type" as a named Python class/type object. I think it's useful here to also show that the general concept of a "static type" isn't limited to nominal typing. I've reworded this a bit to emphasize it less, let me know if it still seems out of place to you.

The subtype relation
--------------------

A static type ``B`` is a **subtype** of another static type ``A`` if and only
if the set of values represented by ``B`` is a subset of the set of values
represented by ``A``. Because the subset relation on sets is transitive and
reflexive, the subtype relation is also transitive (if ``C`` is a subtype of
``B`` and ``B`` is a subtype of ``A``, then ``C`` is a subtype of ``A``) and
reflexive (``A`` is always a subtype of ``A``).

The **supertype** relation is the inverse of subtype: ``A`` is a supertype of
``B`` if and only if ``B`` is a subtype of ``A``; or equivalently, if and only
if the set of values represented by ``A`` is a superset of the values
represented by ``B``. The supertype relation is also transitive and reflexive.

We also define an **equivalence** relation on static types: the types ``A`` and
``B`` are equivalent (or "the same type") if and only if ``A`` is a subtype
of ``B`` and ``B`` is a subtype of ``A``. This means that the set of values
represented by ``A`` is both a superset and a subset of the values represented
by ``B``, meaning ``A`` and ``B`` must represent the same set of values.

> We may describe a type ``B`` as "narrower" than a type ``A`` (or as a "strict

Author: In set terminology, both "strict subset" and "proper subset" are used synonomously. I'm having a hard time finding references for either term used with types, but I don't have any objection to switching to "proper subtype."

subtype" of ``A``) if ``B`` is a subtype of ``A`` and ``B`` is not equivalent
to ``A``.

The consistency relation
------------------------

Since :ref:`Any` represents an unknown static type, it does not represent any

> known single set of values, and thus it is not in the domain of the subtype,

Reviewer 5: This phrasing may not be the best and I'm not sure how to phrase this well, but I do think that the point is actually correct.

The type `list[object]` is *exactly* the set of all possible `list` objects (and is a fully static type that permits exactly the operations that are legal on this entire set).

The type `list[Any]`, on the other hand, represents a list of objects each of whose type is unknown, and so under gradual typing rules we permit each element to be used in arbitrary operations (by way of materializing as in the next paragraph).

Maybe actually discussing this example would clarify things a bit. Using examples may not feel very spec-y, but there's a reason why even math textbooks usually intermingle their abstract definitions with concrete examples.

> known single set of values, and thus it is not in the domain of the subtype,

Reviewer 6: A concise way to think of `list[Any]` is that it is an arbitrary set of list objects, not necessarily all.

(And `list[object]` is subtle because `list` is an invariant generic.  It's the set of all lists that allow inserting `object`s.)

> known single set of values, and thus it is not in the domain of the subtype,

Reviewer 5: I think the question of whether `list[object]` represents all possible lists is related to the other concern I had (not about this change but about typing spec and docs in general) that I want to be sure we don't use language implying a *value* has a type.

Any list value certainly can be a valid `list[object]`, but variance is a good example of why values don't actually have types, because what type operations are allowed is a question not only about the value but about all possible references to it. This is why static types in general only apply to terms, not to values.

So I guess here I was getting sloppy about my language in exactly the way I worry will leak into the spec and docs :)

> known single set of values, and thus it is not in the domain of the subtype,

Author: I would approach this last point slightly differently. I think for a model of "types as sets of values" to work, we need to include the type signatures of methods in our concept of what a "value" is. So a list value with a method `append(self, obj: int) -> None` is distinct from a list value with a method typed `append(self, obj: object) -> None`.

> known single set of values, and thus it is not in the domain of the subtype,

Reviewer 6: Oh, I see what you were saying :)

You have to have a different notion of **value**.  A value is a python object tagged with a fully static type.  (And an unspecialized generic like `list` is not a type.)  So the set of lists tagged with `list[object]` and the set tagged with `list[int]` are really distinct sets of values.

If the type system inserted runtime checked casts, it would use this tag.  But CPython doesn't do that so we can have an implementation that erases the type tags.

> known single set of values, and thus it is not in the domain of the subtype,

Author: I added the example `tuple[int, Any]` above to try to help clarify this distinction. I think a deeper discussion of the generalization to invariant and contravariant generics probably doesn't belong here, but rather in the generics chapter.

Resolving this thread for now, feel free to reopen if you think there's more that needs to be addressed here.

supertype, or equivalence relations on static types described above. (We do

> find it convenient to say, however, that ``Any`` is both subtype and supertype

Author: I agree that this sentence is awkward. Many gradual type systems in the
literature do define Any as a subtype of itself. It makes simplification of
union types more straightforward to define. If we don't allow Any to be a
subtype of itself, then the union type `int | Any | Any` can't be simplified to
`int | Any` under the rule that we can simplify subtypes out of unions. (We
can't simplify unions based on assignability, or else we get e.g. `int | Any`
simplifying to `Any`, which we don't want.)

I think a clearer option here might be to say explicitly in our definition of
unions that `Any | Any` is equivalent to `Any`, because the union of two
unknown sets of objects is an unknown set of objects.

This makes more intuitive sense than saying `Any <: Any`, since we can't really
say that one unknown set of objects is a subset of another unknown set of
objects.

@superbobry reminded me in an earlier comment to consider the case of union
simplification, which led me to add this sentence. @superbobry, do you think it
is actually important that we define `Any <: Any`, or can we instead just
define `Any | Any == Any`?

> find it convenient to say, however, that ``Any`` is both subtype and supertype

Reviewer 7: I cannot think of another case where it is important that `Any <: Any`, but maybe @kmillikin has an example.

> find it convenient to say, however, that ``Any`` is both subtype and supertype

Author: The latest version no longer says that `Any <: Any`, but instead says directly that `Any | Any` is `Any`.

of -- that is, equivalent to -- only itself. This can allow us to simplify

> redundant multiple occurrences of ``Any`` out of more complex types.)

Author: Yes, this is about unions. I was avoiding discussing them specifically since they hadn't been introduced yet, but I agree this was too vague and confusing.

> To relate gradual types more generally, we define a **materialization**

Author: I also think it's a useful and intuitive concept! It was first introduced in
the literature (as far as I'm aware) in the 2019 paper "Gradual Typing: A New
Perspective." (https://www.irif.fr/~gc/papers/popl19.pdf)

I'm not sure how I can introduce it earlier and use it to define gradual types,
though. Since materialization is the process of replacing Any in a type with a
more static type, that seems to depend on first understanding that some types
can contain Any as a sub-part. If you're interested in this idea, maybe you can
suggest in a bit more detail how this presentation would look?

> To relate gradual types more generally, we define a **materialization**

Reviewer: I think your updated wording in the earlier section eliminates the need to introduce this concept earlier, so I withdraw this suggestion.

relation. The intuition for materialization is that it transforms a "more
dynamic" type to a "more static" type. Given a gradual type ``A``, if we
replace zero or more occurrences of ``Any`` in ``A`` with some gradual type
(which can be different for each occurrence of ``Any``), the resulting gradual
type ``B`` is a materialization of ``A``.

For instance, ``tuple[int, str]`` (a static type) and ``tuple[Any, str]`` (a
gradual type) are both materializations of ``tuple[Any, Any]``. ``tuple[int,
str]`` is also a materialization of ``tuple[Any, str]``.

If ``B`` is a materialization of ``A``, we can say that ``B`` is a "more
static" type than ``A``, and ``A`` is a "more dynamic" type than ``B``.

The materialization relation is both transitive and reflexive, so it defines a
preorder on gradual types.

We also define a **consistency** relation on gradual types.

> A static type ``A`` is consistent with another static type ``B`` if and only if

Author: Yeah, I think that's a loose/colloquial usage. It's a reasonable usage given the informal meaning of the term, but the meaning presented here is the one used in the literature, and I don't think there's a sufficiently strong reason to use something different. I don't think there will be very many places in the spec where we need the consistency relation at all. TypeVar constraints will be one.

> A static type ``A`` is consistent with another static type ``B`` if and only if

Reviewer 8: In some literature, such as Abstracting Gradual Types, the equivalence relation is lifted to the consistency relation. Are we saying that consistency can be thought of as consistent subtyping, since in practice, consistency doesn't actually come up that much since we are usually considering systems with subtyping? Right now the actual definitions look similar to what's in the literature.

> A static type ``A`` is consistent with another static type ``B`` if and only if

Reviewer 6: No.  According to AGT, lifting static typing **equivalence** gives gradual typing **consistency**; and lifting static typing **subtyping** gives gradual typing **consistent subtyping**.

The idea of AGT is that you design a fully statically typed language, which is "easy".  And then you get gradually typed equivalents of the static typing constructs "for free" by using their approach.

> A static type ``A`` is consistent with another static type ``B`` if and only if

Reviewer 6: PEP-483 loosely defines an "is-consistent-with" relation but it means consistent subtyping.  That has understandably led people to use the term "consistent" when they are talking about consistent subtyping, which could be confusing.

We almost don't need consistency at all, so it hasn't caused us any real problems to use the wrong term for consistent subtyping (other than constantly annoying me :) ).

> A static type ``A`` is consistent with another static type ``B`` if and only if

Reviewer 8: Makes sense! I also would immediately think of the standard consistency relation (instead of consistent subtyping) when seeing the consistency term discussed above. So the description above looks right to me. But I could also be one of a few people!

          e = [e]
      ...

A union type ``T1 | T2``, where ``T1`` and ``T2`` are fully static types,
represents the set of values formed by the union of all the sets of values
represented by ``T1`` and ``T2``. Thus, by the definition of the supertype
relation, the union ``T1 | T2`` is a supertype of both ``T1`` and ``T2``. Via
materialization, the gradual types ``S1`` and ``S2`` are both assignable to a
gradual union type ``S1 | S2``.

If ``B`` is a subtype of ``A``, ``B | A`` is equivalent to ``A``.

> This rule applies only to subtypes, not assignable-to. The union ``T | Any`` is

Reviewer: I'm so glad that we're finally clarifying this point — and that the formulation is well justified by the theoretical framework.

          e = [e]
      ...

A union type ``T1 | T2``, where ``T1`` and ``T2`` are fully static types,
represents the set of values formed by the union of the sets of values

> represented by ``T1`` and ``T2``. Thus, by the definition of the supertype

Reviewer 2: ```suggestion
represented by ``T1`` and ``T2`` respectively. Thus, by the definition of the supertype
```

relation, the union ``T1 | T2`` is a supertype of both ``T1`` and ``T2``. Via
materialization, the gradual types ``S1`` and ``S2`` are both assignable to a

> gradual union type ``S1 | S2``.

Reviewer 3: This sentence doesn't make much sense to me. It might be clearer if you say more explicitly what S1 | S2 means if the two types are not fully static.

Author: Does this added sentence help to clarify?

> A gradual union type ``S1 | S2``, where ``S1`` and ``S2`` are gradual types,
represents all possible sets of values that could be formed by union of the
possible sets of values represented by materializations of ``S1`` and ``S2``.

Author: Although actually I think this explanation might be more directly related to the sentence you asked about:

> Via materialization, the gradual types ``S1`` and ``S2`` are both assignable to
a gradual union type ``S1 | S2``. For any materialization of ``S1`` to ``T1``
and ``S2`` to ``T2``, ``S1 | S2`` can likewise be materialized to ``T1 | T2``,
where ``T1`` and ``T2`` are subtypes of ``T1 | T2``.

Reviewer 3: The current version of this section reads much better, yes. Thanks!

types, represents the set of values formed by the union of the sets of values
represented by ``T1`` and ``T2``, respectively. Thus, by the definition of the
supertype relation, the union ``T1 | T2`` is a supertype of both ``T1`` and
``T2``, and ``T1`` and ``T2`` are both subtypes of ``T1 | T2``.

A gradual union type ``S1 | S2``, where ``S1`` and ``S2`` are gradual types,
represents all possible sets of values that could be formed by union of the
possible sets of values represented by materializations of ``S1`` and ``S2``,
respectively.

For any materialization of ``S1`` to ``T1`` and ``S2`` to ``T2``, ``S1 | S2``
can likewise be materialized to ``T1 | T2``. Thus, the gradual types ``S1`` and
``S2`` are both assignable to the gradual union type ``S1 | S2``.

If ``B`` is a subtype of ``A``, ``B | A`` is equivalent to ``A``.

This rule applies only to subtypes, not assignable-to. The union ``T | Any`` is
not reducible to a simpler form. It represents an unknown static type with
lower bound ``T``. That is, it represents an unknown set of objects which may
be as large as ``object``, or as small as ``T``, but no smaller.

> As a special case, the union ``Any | Any`` can be simplified to ``Any``: the

Author: Hmm, I think you're right, but this gets a bit tricky. We haven't defined what it means for two gradual types to be equivalent (the same gradual type, as opposed to just consistent.) Consistent-with isn't sufficient to simplify types out of a union. I'll give some thought to how to best define this.

> As a special case, the union ``Any | Any`` can be simplified to ``Any``: the

Author: Added this. Thanks for the note!

          e = [e]
      ...

A fully static union type ``T1 | T2``, where ``T1`` and ``T2`` are fully static
types, represents the set of values formed by the union of the sets of values
represented by ``T1`` and ``T2``, respectively. Thus, by the definition of the
supertype relation, the union ``T1 | T2`` is a supertype of both ``T1`` and
``T2``, and ``T1`` and ``T2`` are both subtypes of ``T1 | T2``.

> A gradual union type ``S1 | S2``, where ``S1`` and ``S2`` are gradual types,

Reviewer 9: ```suggestion
A gradual union type ``S1 | S2``, where ``S1`` or ``S2`` are gradual types,
```
is this right? if a fully static type is a gradual type, we should mention that when defining gradual type

Reviewer 2: I think if you read Carl's PR closely, it's clear that all fully static types are gradual types: all Python types are gradual, so the set of fully static types is a strict subset of the set of gradual types (which is the same as the set of all types). This is indicated in passages in `concepts.rst` such as:

>     A **gradual type** can be a fully static type, :ref:`Any` itself, or a type
>     that contains a gradual form as a sub-part.

and in passages in `glossary.rst` such as:

>      Types in the Python type system are "gradual". A gradual type may be a
>      :term:`fully static type`, or it may be :ref:`Any`, or a type that
>      contains ``Any`` or another :term:`gradual form`.

Nonetheless, I agree that this is quite a subtle point and one that I was confused about in a prior version of the PR (see https://github.com/python/typing/pull/1743#discussion_r1623286059). I think it's easy to misread as although the language of the document is clear, the structure of the document overall often pits the two concepts against each other in contrast (first talking about one category, then talking about how it's different for the other category), such that a reader can easily misunderstand the two categories as being mutually exclusive. It's also not helped by the fact that we don't seem to have a precise term for "gradual types that include at least one gradual form as a subelement, such that the materialization from gradual type to fully static type transforms the type in some way".

Here are some possible edits to the PR that might clarify this somewhat for readers:

```diff
--- a/docs/spec/concepts.rst
+++ b/docs/spec/concepts.rst
@@ -54,6 +54,11 @@ as **fully static types**.
 A **gradual type** can be a fully static type, :ref:`Any` itself, or a type
 that contains a gradual form as a sub-part.
 
+The categories "fully static" and "gradual" are not mutually exclusive.
+Python's type system is gradual; all Python types are gradual types, including
+those types which are fully static. The set of fully static types is a strict
+subset of the set of gradual types.
+
 Fully static types
 ~~~~~~~~~~~~~~~~~~
 
@@ -70,11 +75,11 @@ If an object ``v`` is a member of the set of objects denoted by a fully static
 type ``T``, we can say that ``v`` is a "member of" the type ``T``, or ``v``
 "inhabits" ``T``.
 
-Gradual types
-~~~~~~~~~~~~~
+Gradual forms and gradual types
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-:ref:`Any` represents an unknown static type. It denotes some unknown set of
-runtime values.
+:ref:`Any` is a gradual form that represents an unknown static type. It denotes
+some unknown set of runtime values.
 
 This may appear similar to the fully static type ``object``, which represents
 the set of all Python objects, but it is quite different.
diff --git a/docs/spec/glossary.rst b/docs/spec/glossary.rst
index 41bb2b5..1e9bae2 100644
--- a/docs/spec/glossary.rst
+++ b/docs/spec/glossary.rst
@@ -55,8 +55,9 @@ This section defines a few terms that may be used elsewhere in the specification
    fully static type
       A type is "fully static" if it does not contain any :term:`gradual form`.
       Fully static types represent a set of possible runtime values. Fully
-      static types participate in the :term:`subtype` relation. See
-      :ref:`type-system-concepts`.
+      static types participate in the :term:`subtype` relation. Just because a
+      type is fully static does not mean that it is not gradual: all types in
+      Python's type system are gradual types. See :ref:`type-system-concepts`.
 
    gradual form
       A gradual form is a :term:`type expression` which makes the type it is
@@ -68,7 +69,7 @@ This section defines a few terms that may be used elsewhere in the specification
       other :ref:`tuple <tuples>` types).
 
    gradual type
-      Types in the Python type system are "gradual". A gradual type may be a
+      All types in the Python type system are "gradual". A gradual type may be a
       :term:`fully static type`, or it may be :ref:`Any`, or a type that
       contains ``Any`` or another :term:`gradual form`. A gradual type does not
       necessarily represent a single set of possible runtime values; instead it
```

Reviewer 9: Ah thanks, I'd missed that. When I tried to find an answer, I went and looked at the "Gradual types" section, which doesn't mention this (although the "Fully static and gradual types" supersection does).

I think your changes suggested changes would have avoided my confusion, although the "Gradual types" section still has some wording that implies "gradual type" as a shorthand for "type containing a gradual form"
